### PR TITLE
TypeHintSniff: fix return evaluation issues

### LINKED ISSUE #40: Problem in return point detection
There are 2 issues with `TypeHintSniff` when detecting return points.

## First issue

The first problem regards a code like this:

```php
<?php
function returnZero() : int
{
    return 0;
}
```

That even if fine, raises an error like:

```
---------------------------------------------------------------------------------------------
WARNING | Return type with no return (NeutronStandard.Functions.TypeHint.UnusedReturnType)
---------------------------------------------------------------------------------------------
```

This is not caused in the `TypeHintSniff.php` file itself, but in `SniffHelpers::isReturnValueVoid()` where a there's the line:

```php
return ! (
    $returnValue
    && $returnValue['content']   // <-- this is problematic 
    && $returnValue['code'] !== 'PHPCS_T_SEMICOLON'
);
```

this is problematic because when `$returnValue['content']` is `'0'` or `''` (empty string) the return value is still non-void but `$returnValue['content']` will evaluate as `false`.

Considering that when the token is not a whitespace and is not a semicolon it can be safely considered as non-void there's no need to check the "content" attribute.

---

## Second issue

The second problem regards a code like this:

```php
function test(bool $condition): int
{
   if ($condition) {
      return 1;
   }

   return;
}
```

Which is clearly wrong and could cause a fatal error, but is not recognized as problematic, because currently if any non-void return point is found and the function as a non-void return type no error is raised.

This could be fixed counting all the void and non-void return token and then apply proper logic based on the count found.

---

## One more thing

Related to this, another issue IMO is that `UnusedReturnType` and `IncorrectVoidReturnType` should be errors, and not warnings, because executing the code where those are detected will cause fatal errors.

### PR DESCRIPTION
**Pull request by gmazzap**
Fixes #40

### CODE REVIEW

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php [hunk 1]
```diff
@@ -30,53 +30,61 @@ private function checkForMissingArgumentHints(File $phpcsFile, $stackPtr) {
...
-		$endOfFunctionPtr = $this->helper->getEndOfFunctionPtr($phpcsFile, $stackPtr);
-		$startOfFunctionPtr = $this->helper->getStartOfFunctionPtr($phpcsFile, $stackPtr);
-		$returnTypePtr = $this->helper->getNextReturnTypePtr($phpcsFile, $stackPtr);
+		$endOfFunctionPtr = $helper->getEndOfFunctionPtr($phpcsFile, $stackPtr);
+		$startOfFunctionPtr = $helper->getStartOfFunctionPtr($phpcsFile, $stackPtr);
+		$returnTypePtr = $helper->getNextReturnTypePtr($phpcsFile, $stackPtr);
 		$returnType = $tokens[$returnTypePtr];
 
-		$foundReturn = false;
+		$nonVoidReturn = 0;
```
Comment (sirbrillig) [inline NeutronStandard/Sniffs/Functions/TypeHintSniff.php]: Could we call these `nonVoidReturnCount` and `voidReturnCount` just to make it clear that they hold an integer?

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php [hunk 2]
```diff
@@ -30,53 +30,61 @@ private function checkForMissingArgumentHints(File $phpcsFile, $stackPtr) {
...
+			if (empty($scopeClosers) && $token['code'] === T_RETURN) {
+				$helper->isReturnValueVoid($phpcsFile, $ptr) ? $voidReturn++ : $nonVoidReturn++;
 			}
 		}
 
-		if (! $foundReturn && $returnTypePtr && $returnType['content'] !== 'void') {
-			$error = 'Return type with no return';
-			$phpcsFile->addWarning($error, $stackPtr, 'UnusedReturnType');
+		if ($returnTypePtr && $returnType['content'] !== 'void') {
+			if (!$nonVoidReturn) {
```
Comment (sirbrillig) [inline NeutronStandard/Sniffs/Functions/TypeHintSniff.php]: The logic in this block is great, but I find it a little hard to follow. Mostly I think it's because of the double-negative "if not non-void return" and because there's two sub-conditions. Also because the first condition includes an early return. I wonder if there's a way to decrease the nesting and increase the readability without being overly verbose. 🤔I'm still thinking about this but maybe you'll have better ideas than me.

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php [hunk 3]
```diff
@@ -30,53 +30,61 @@ private function checkForMissingArgumentHints(File $phpcsFile, $stackPtr) {
...
+				$error = 'Return type with no return';
+				$phpcsFile->addError($error, $stackPtr, 'UnusedReturnType');
+
+				return;
+			}
+			if ($voidReturn) {
```
Comment (sirbrillig) [inline NeutronStandard/Sniffs/Functions/TypeHintSniff.php]: Since these are integers, could we explicitly make the comparisons integer comparisons rather than a truthy check? So this would be `if ($voidReturn > 0)` and the ones below would be `if ($nonVoidReturn > 0 && ...`. I know that a truthy check is the same result here, but I worry that in the future it might trick me into accidentally assuming one of these is a boolean, causing a similar problem to the one this PR is designed to solve. 😁

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php [hunk 4]
```diff
@@ -12,12 +12,12 @@ public function register() {
...
-		$this->helper = new SniffHelpers();
-		$this->checkForMissingArgumentHints($phpcsFile, $stackPtr);
-		$this->checkForMissingReturnHints($phpcsFile, $stackPtr);
+		$helper = new SniffHelpers();
+		$this->checkForMissingArgumentHints($phpcsFile, $stackPtr, $helper);
+		$this->checkForMissingReturnHints($phpcsFile, $stackPtr, $helper);
 	}
 
-	private function checkForMissingArgumentHints(File $phpcsFile, $stackPtr) {
+	private function checkForMissingArgumentHints(File $phpcsFile, $stackPtr, SniffHelpers $helper) {
```
Comment (sirbrillig) [inline NeutronStandard/Sniffs/Functions/TypeHintSniff.php:20]: This is a great idea. I should have done this rather than keeping `helper` as state. Thanks!

File: NeutronStandard/SniffHelpers.php [hunk 5]
```diff
@@ -55,7 +55,7 @@ public function isReturnValueVoid(File $phpcsFile, $stackPtr) {
 			return false;
 		}
 		$returnValue = $this->getNextNonWhitespace($phpcsFile, $stackPtr);
-		return ! ($returnValue && $returnValue['content'] && $returnValue['code'] !== 'PHPCS_T_SEMICOLON');
+		return ! $returnValue || $returnValue['code'] === 'PHPCS_T_SEMICOLON';
```
Comment (sirbrillig) [inline NeutronStandard/SniffHelpers.php:58]: Nice find!

**Comment by sirbrillig** [comment]
Thanks for this work!

> Restore a wrong-styled line in fixture to don't break tests

Yeah, some of the fixtures have weird formatting intentionally to make sure we can handle them correctly, just in case! 😄

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php (hunk 4 above)
Comment (gmazzap) [reply NeutronStandard/Sniffs/Functions/TypeHintSniff.php:20]: Did it because you wer using a non-defined helper property which was marked as error in my IDE :)

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php (hunk 1 above)
Comment (gmazzap) [reply NeutronStandard/Sniffs/Functions/TypeHintSniff.php]: Sure

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php (hunk 2 above)
Comment (gmazzap) [reply NeutronStandard/Sniffs/Functions/TypeHintSniff.php]: I could try to make it simple

File: NeutronStandard/Sniffs/Functions/TypeHintSniff.php (hunk 3 above)
Comment (gmazzap) [reply NeutronStandard/Sniffs/Functions/TypeHintSniff.php]: If they are called `$voidReturnCount` as you suggested should not be wrongly asumed as boolean, but if you want to have that, it's easy and does not hurt.

**Review by sirbrillig — APPROVED** [review]
Looks great! Thanks!

**Comment by gmazzap** [comment]
Thanks for merging @sirbrillig! Any plan to release a new version with last changes?

**Comment by sirbrillig** [comment]
@gmazzap Yup! I'll do one today. I was just waiting 24 hours to see if there were any more issues. ;-)
